packages/techdocs-common/src/stages/publish/awsS3.ts

```diff
@@ -217,27 +217,23 @@ export class AwsS3Publish implements PublisherBase {
       const fileExtension = path.extname(filePath);
       const responseHeaders = getHeadersForFileExtension(fileExtension);
 
+      // Pipe file chunks directly from storage to client.
       this.storageClient
         .getObject({ Bucket: this.bucketName, Key: filePath })
-        .then(async object => {
-          const fileContent = await streamToBuffer(object.Body as Readable);
-          if (!fileContent) {
-            throw new Error(`Unable to parse the file ${filePath}.`);
-          }
-
-          // Inject response headers
-          for (const [headerKey, headerValue] of Object.entries(
-            responseHeaders,
-          )) {
-            res.setHeader(headerKey, headerValue);
-          }
-
-          res.send(fileContent);
+        .createReadStream()
+        .on('pipe', () => {
+          res.writeHead(200, responseHeaders);
         })
-        .catch(err => {
+        .on('error', (err: { message: string }) => {
           this.logger.warn(err.message);
-          res.status(404).send(err.message);
-        });
+          // Send a 404 with a meaningful message if possible.
+          if (!res.headersSent) {
+            res.status(404).send(err.message);
```

Comment:
Better pass the error to `next` instead, letting the error handler take care of response representation, logging, etc.

Otherwise use `.json` instead of send here, or set an explicit content-type